devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,576 @@
+	response(requests)
+}
+
+def parameterSetting() {
+	def lockmode = null
+	def valid_lock = false
+
+	log.debug "lock : $settings.lock"
+	if (settings.lock == "Yes") {
+		lockmode = 0x01
+		valid_lock = true
+	} else if (settings.lock == "No") {
+		lockmode = 0x00
+		valid_lock = true
+	}
+
+	if (valid_lock) {
+		log.debug "lock valid"
+		delayBetween([
+				zigbee.writeAttribute(THERMOSTAT_UI_CONFIG_CLUSTER, ATTRIBUTE_KEYPAD_LOCKOUT, DataType.ENUM8, lockmode),
+				poll()
+			], 200)
+	} else {
+		log.debug "nothing valid"
+	}
+}
+
+def parse(String description) {
+	log.debug "Parse description $description"
+	def map = [:]
+
+	// If the user installed with an old DTH version, update so that the new mobile client will work
+	if (!device.currentValue("supportedThermostatModes")) {
+		configureSupportedRanges()
+	}
+	// Existing installations need the temperatureAlarm state initialized
+	if (device.currentValue("temperatureAlarm") == null) {
+		sendEvent(name: "temperatureAlarm", value: "cleared", displayed: false)
+	}
+
+	if (description?.startsWith("read attr -")) {
+		def descMap = zigbee.parseDescriptionAsMap(description)
+		log.debug "Desc Map: $descMap"
+		if (descMap.clusterInt == THERMOSTAT_CLUSTER) {
+			if (descMap.attrInt == ATTRIBUTE_LOCAL_TEMP) {
+				map.name = "temperature"
+				map.unit = getTemperatureScale()
+				map.value = getTemperature(descMap.value)
+
+				if (descMap.value == "7ffd") {		// 0x7FFD
+					map.name = "temperatureAlarm"
+					map.value = "freeze"
+					map.unit = ""
+				} else if (descMap.value == "7fff") {	// 0x7FFF
+					map.name = "temperatureAlarm"
+					map.value = "heat"
+					map.unit = ""
+				} else if (descMap.value == "8000") {	// 0x8000
+					map.name = null
+					map.value = null
+					map.descriptionText = "Received a temperature error"
+				} else if (descMap.value > "8000") {
+					map.value = -(Math.round(2*(655.36 - map.value))/2)
+				}
+
+				if (device.currentValue("temperatureAlarm") != "cleared" && map.name == "temperature") {
+					sendEvent(name: "temperatureAlarm", value: "cleared")
+				}
+			} else if (descMap.attrInt == ATTRIBUTE_HEAT_SETPOINT) {
+				if (descMap.value != "8000") {		//0x8000
+					log.debug "HEATING SETPOINT"
+					map.name = "heatingSetpoint"
+					map.value = getTemperature(descMap.value)
+					map.data = [heatingSetpointRange: heatingSetpointRange]
+				}
+			} else if (descMap.attrInt == ATTRIBUTE_SYSTEM_MODE) {
+				log.debug "MODE - ${descMap.value}"
+				def value = modeMap[descMap.value]
+
+				// If we receive an off here then we are off
+				// Else we will determine the real mode in the mfg specific packet so store this
+				if (value == "off") {
+					map.name = "thermostatMode"
+					map.value = value
+					map.data = [supportedThermostatModes: supportedThermostatModes]
+				} else {
+					state.storedSystemMode = value
+				}
+				// Right now this doesn't seem to happen -- regardless of the size field the value seems to be two bytes
```

Comment:
In these cases you can look to see if there is anything in the "additionalAttrs" key as if more than one attribute value is reported we strip it out.  However, I'd have to see exactly how they are reporting it to see if our current parsing can handle it.

Reply:
Yeah, unfortunately there was nothing there... :/ I'll dig into that a little more, but so far it seems to be working as is.